Add guideline on conventional pattern for code-based user interface (#22)
scipp/essreduce: docs/user-guide/reduction-workflow-guidelines.md

  E.g., `sc.scalar(2, unit='m')` and `sc.scalar(2.0, unit='m')` should be usable interchangeably.
  This can also apply to arrays, for instance, `sc.linspace` and `sc.arange` should be interchangeable but the latter may result in integers while the former typically produces floats.

> ### C.5: Use a conventional pattern minimizing programming jargon, favor fewer and concepts

Reviewer: Not sure how you mean this sentence. Is it
```suggestion
### C.5: Use a conventional pattern minimizing programming jargon, favor fewer concepts
```
or
```suggestion
### C.5: Use a conventional pattern minimizing programming jargon and concepts and favor fewer concepts
```
?

In any case, I would be more specific:
```suggestion
### C.5: Use a fixed pattern for creating, manipulating, and running workflows
```

**Reason**

> - Terms such as provider or pipeline mean nothing to users, using their language instead reduces cognitive load.

Reviewer: ```suggestion
- Terms such as provider or pipeline mean nothing to users, using their natural language and instead reduces cognitive load.
```

Author: Their natural language is English, I was referring to how they are used to referring to things in their work/research.

Author: That does not parse grammatically.

Reviewer: True. Let's be more explicit:
```
Using terms such as provider or pipeline increases cognitive load for scientific users.
The reason is that those terms are unfamiliar to users based on the scientific domain language they are used to.
```

Author: ```suggestion
- Using terms such as provider or pipeline increases cognitive load for scientific users.
  The reason is that those terms are unfamiliar to users based on the scientific domain language they are used to.
```

- Manipulating multiple concepts such as (1) a dict of parameters, (2) a `sciline.Pipeline`, and (3) a `sciline.TaskGraph` is confusing, especially for non-programmers.

**Notes**

- Add one or more `make_workflow` function(s) that returns a `sciline.Pipeline` object, configured with default providers and parameters.
  - Prefix with `make` and suffix with `workflow`, qualifiers can be added in between, for example `make_abc_workflow`.
- Avoid creating parameter dicts in notebooks, set parameters on the workflow object directly.
- Avoid calling `workflow.get` (which would return a `sciline.TaskGraph`), instead call `workflow.compute` and `workflow.visualize`, even if it means listing the result and building the task graph multiple times.

**Example**

```python
from ess import loki

> workflow = loki.make_workflow()

Author: I am tempted to change this to a camel-case function name (`loki.LokiWorkflow()` or sth. like that). Wouldn't it make it easier to switch to a class later, without having to touch all the notebook?

Author: @jl-wynen Thoughts on this one?

Reviewer: You don't have to touch the notebooks. It would still be possible to have a function that instantiates the class. Then there would be two ways of doing the same thing which would not be great. But it wouldn't be a big issue either, IMHO.

Author: Ok, then let us leave it as it is.
